### PR TITLE
DPL-258 new barcode format

### PR DESCRIPTION
**Pull request by BenTopping**
Closes #

Changes proposed in this pull request:

* Added new sequence for sequencescape plates
* Changes to barcode conversion and formatting
* Includes debug/logging changes

### CODE REVIEW

File: baracoda/operations.py [hunk 1]
```diff
@@ -167,3 +173,41 @@ def __set_prefix_item(self):
             prefix item or None if prefix does not exist
         """
         self.prefix_item = get_prefix_item(self.prefix)
+
+
+class ChildBarcodeOperations:
```
Comment (emrojo) [inline baracoda/operations.py:178]: This class only has one method and it is static, which is equivalent to having a function inside a module.

File: baracoda/barcodes.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 @bp.post("/barcodes_group/<prefix>/new")  # type: ignore
 def get_new_barcode_group(prefix: str) -> Tuple[Any, int]:
     try:
+        logger.debug(f"Creating a barcode group for '{ prefix }'")
```
Comment (stevieing) [inline baracoda/barcodes.py:20]: There seem to be quite a few logger debugs. Are these necessary? If we start using this as oour primary source for creating barcodes will it just create a lot of noise?

File: baracoda/child_barcodes.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import logging
```
Comment (stevieing) [inline baracoda/child_barcodes.py:1]: Wondering if a bit more documentation would be nice? Just to explain why this is here and what it is doing.

File: baracoda/formats.py [hunk 4]
```diff
@@ -3,14 +3,39 @@
 logger = logging.getLogger(__name__)
 
 
-class HeronFormatter:
-    def __init__(self, prefix: str, convert: bool = True):
+class FormatterInterface:
```
Comment (stevieing) [inline baracoda/formats.py:6]: Nice.

File: baracoda/operations.py [hunk 5]
```diff
@@ -1,18 +1,24 @@
 import logging
```
Comment (stevieing) [inline baracoda/operations.py:1]: Again although the code is readable it is not quite clear exactly what is going on. Maybe some more general documentation about the reasoning. Baracode is now our primary way of producing barcodes and could be used by others.

File: tests/data/fixture_data.py [hunk 6]
```diff
@@ -1,34 +1,40 @@
 from typing import Dict, List, Union
+from baracoda.formats import HeronCogUkIdFormatter, GenericBarcodeFormatter
 
-PREFIXES: List[Dict[str, Union[str, bool]]] = [
+PREFIXES: List[Dict[str, Union[str, object]]] = [
     {
         "prefix": "SANG",
         "sequence_name": "heron",
-        "convert": True,
+        "formatter_class": HeronCogUkIdFormatter,
```
Comment (stevieing) [inline tests/data/fixture_data.py:8]: Liking the idea of formatters. When I initially did this we only had the one format so this is a natural progression and makes Baracode more flexible without adding complexity. Well done.

File: tests/test_child_barcodes.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+    response = client.post(
+        "/child-barcodes/test/new",
+        data=json.dumps({"barcode": "test", "count": 0}),
+        headers={"Content-Type": "application/json"},
+    )
+    assert response.json == {"errors": ["InvalidCountError"]}
+    assert response.status_code == HTTPStatus.UNPROCESSABLE_ENTITY
+
+
+# def test_child_barcodes_use_previous_parent_when_present_allowed(app):
```
Comment (stevieing) [inline tests/test_child_barcodes.py:283]: Tests no longer needed? Can these comments be removed.

**Review by stevieing — CHANGES_REQUESTED** [review]
Pretty good. Readable and looks like you have created a situation where it can be adapted further. My only comments are about documentation. It might be good to provide some more in depth explanations of what is going on. I think Baracoda has a good future and it would help if other teams within the organisation want to use it.

**Comment by emrojo** [comment]
Added workflow of barcodes in lucid chart:

File: baracoda/barcodes.py (hunk 2 above)
Comment (emrojo) [reply baracoda/barcodes.py:20]: We have a setting to specify the level of logs that will be enable. I've seen in code it was set up to DEBUG, but I'll change it to ERROR (only log when there is an error or something more critical). This way it won't log this lines unless we start the application in DEBUG mode, so they can be useful to debug in local if we need. I'll do the change in config.

Comment (emrojo) [reply baracoda/barcodes.py:38]: Changed log level

File: baracoda/child_barcodes.py (hunk 3 above)
Comment (emrojo) [reply baracoda/child_barcodes.py:1]: Ok, added more documentation

File: README.md [hunk 8]
```diff
@@ -151,11 +167,105 @@ The following routes are available from this service:
...
+sequence.
+- ```formatter_class```: Defines the class that will generate the string that represents
+a new barcode by using the prefix and the new value obtained from the sequence.
+If we want to support a new formatter class we have to provide a class that implements
+the interface ```baracoda.formats.FormatterInterface```.
+- ```enableChildrenCreation```: Defines if the prefix has enabled the children creation.
+If true, the prefix will support creating barcodes based on a parent barcode
+If it is False, the prefix will reject any children creation request for that prefix.
+
+### Children barcode creation
```
Comment (stevieing) [inline README.md:203]: This would be better as Child barcode creation

File: README.md [hunk 9]
```diff
@@ -6,6 +6,20 @@
 
 Barcode generation using postgres sequences and pre-defined prefixes.
```
Comment (stevieing) [inline README.md:7]: Really good

**Review by stevieing — APPROVED** [review]
All good. Like the additional documentation.

**Pull request merged**